feat: node mempool limiting

natebeauregard opened the pull request:
Closes https://github.com/AztecProtocol/aztec-packages/issues/12879

[Design doc](https://github.com/AztecProtocol/engineering-designs/blob/main/docs/mempools/dd.md)

Currently, there are no limits on the size of the pending tx pool.

This PR adds limits on the total number and cumulative tx size of valid pending txs in the mempool. If either limit is hit, then the lowest priority pending txs will be evicted to make room until both limits are satisfied.

Additionally, whenever a new block is mined, all pending txs will be checked to ensure that they don't share any nullifiers with mined txs, have a sufficient fee payer balance, and have a max block number higher than the mined block.

In the case of a reorg, all pending txs are checked to ensure that their archive root is still valid. If it isn't, then the tx is evicted.

natebeauregard committed:
feat: node mempool limiting

spalladino commented on yarn-project/p2p/src/config.ts:
```diff
@@ -354,6 +366,21 @@ export const p2pConfigMappings: ConfigMappingsType<P2PConfig> = {
       'AuthRegistry, FeeJuice.increase_public_balance, Token.increase_public_balance, FPC.prepare_fee',
   },
+  // TODO: should P2P_MAX_TX_POOL_COUNT/P2P_MAX_TX_POOL_SIZE be configurable or a global constant?
+  // TODO: what should we do if the value is 0? Should we allow unbounded mempools?
```
I'm fine with either unbounded (if it helps testing) or just failing to initialize

spalladino commented on yarn-project/foundation/src/config/env_var.ts:
```diff
@@ -116,6 +116,8 @@ export type EnvVar =
   | 'P2P_TX_POOL_KEEP_PROVEN_FOR'
   | 'P2P_ATTESTATION_POOL_KEEP_FOR'
   | 'P2P_ARCHIVED_TX_LIMIT'
+  | 'P2P_MAX_TX_POOL_COUNT'
+  | 'P2P_MAX_TX_POOL_SIZE'
```
If it helps simplifying things, I'm fine choosing just one of these (probably size)

natebeauregard replied:
Sounds good, I can remove `P2P_MAX_TX_POOL_COUNT`

spalladino commented on yarn-project/p2p/src/mem_pools/tx_pool/aztec_kv_tx_pool.ts:
```diff
@@ -300,4 +342,181 @@ export class AztecKVTxPool implements TxPool {
+  private async evictLowPriorityTxs(
+    newTxHashes: TxHash[],
+  ): Promise<{ numLowPriorityTxsEvicted: number; numNewTxsEvicted: number }> {
+    let pendingTxsByPriority = (await toArray(this.#pendingTxPriorityToHash.valuesAsync())).map(TxHash.fromString);
```
The `toArray` will mean we go through all tx hashes every time this function is called, which could be very often since it's being called from `addTxs`. Let's use the iterator to go lazily over the list.

spalladino commented on yarn-project/p2p/src/mem_pools/tx_pool/aztec_kv_tx_pool.ts:
```diff
@@ -193,14 +232,18 @@ export class AztecKVTxPool implements TxPool {
       );
 
-      this.#metrics.recordAddedObjects(pendingCount, 'pending');
+      const { numLowPriorityTxsEvicted, numNewTxsEvicted } = await this.evictLowPriorityTxs(
+        hashesAndStats.map(({ txHash }) => txHash),
+      );
```
@alexghr this function is reading from LMDB within a write tx. Will it be able to see what was written within this same tx? In other words, if we do:
```
store.write(1)
store.transaction(() => {
  store.write(2)
  value = store.read()
})
```

What will be the value of `value`?

spalladino commented on yarn-project/p2p/src/mem_pools/tx_pool/aztec_kv_tx_pool.ts:
```diff
@@ -300,4 +342,181 @@ export class AztecKVTxPool implements TxPool {
+      }
+
+      // Evict pending txs with an insufficient fee payer balance
+      const db = this.#worldStateSynchronizer.getCommitted();
+      const gasTxValidator = new GasTxValidator(
+        new DatabasePublicStateSource(db),
+        ProtocolContractAddress.FeeJuice,
+        GasFees.empty(),
+      );
```
Let's initialize this out of the tx loop, IIRC getting a world state view is not cheap

spalladino replied:
Also, we need to make sure world state is synchronized to the block we're processing before calling `getCommitted`. Both p2p and world state sync from the archiver at their own pace. If `markAsMined` gets called and reaches this point _before_ world state had time to sync, we'll be seeing the old balances.

spalladino reviewed (commented):
Added a few comments, but looking good!

natebeauregard commented on yarn-project/p2p/package.json:
```diff
@@ -72,6 +72,7 @@
     "@aztec/noir-contracts.js": "workspace:^",
     "@aztec/noir-protocol-circuits-types": "workspace:^",
     "@aztec/protocol-contracts": "workspace:^",
+    "@aztec/sequencer-client": "workspace:^",
```
Yeah I think this is currently broken in my local branch, I still need to look some more into this but may end up just moving `ArchiveCache` to the `p2p` package

natebeauregard replied:
It looks like I was wrong and there wasn't any warning triggered. The build completes without error so I left this in for the time being. However, let me know if you think I should move `ArchiveCache` into the `p2p` package for consistency.

spalladino commented on yarn-project/p2p/src/mem_pools/tx_pool/aztec_kv_tx_pool.ts:
```diff
@@ -300,4 +342,181 @@ export class AztecKVTxPool implements TxPool {
+    const pendingTxs = await Promise.all(
+      pendingTxHashes.map(async hash => ({
+        hash,
+        tx: await this.getTxByHash(hash),
```
Once an initial version of this PR is done, let's benchmark how expensive it is to rehydrate all txs, since we'll be deserializing all txs in the mempool every 36s. It may be worthwhile keeping the hydrated instances in memory for easy access, if it's not too much memory and saves time.

spalladino commented on yarn-project/p2p/src/mem_pools/tx_pool/aztec_kv_tx_pool.ts:
```diff
@@ -300,4 +342,181 @@ export class AztecKVTxPool implements TxPool {
+    const pendingTxs = await Promise.all(
+      pendingTxHashes.map(async hash => ({
+        hash,
+        archiveRoot: await (await this.getTxByHash(hash))?.data.constants.historicalHeader.hash(),
```
Let's also keep an eye on the cost of computing header hashes. We may have to store them to avoid recomputing.

spalladino reviewed (commented):
Added a few notes on perfromance considerations that can be tackled after this PR lands

natebeauregard committed:
chore: add unit tests and update impl bugs

natebeauregard commented:
@spalladino I added unit tests and fixed a few bugs with my original implementation and pushed back up. I believe that all of your comments other than the benchmarking work and performance considerations should be addressed. I went ahead and marked the PR as ready for review. In the meantime, I plan on looking into and comparing performance for:
- rehydrating all txs every iteration or keeping them in memory
- computing header hashes every iteration or storing them
-  checking tx priority before inserting txs into the pool instead of adding and then evicting low priority txs after

natebeauregard committed:
chore: add min gas limit unit tests

natebeauregard committed:
chore: move archive cache to p2p package

spalladino reviewed (commented):
Looking really good! Left just a few comments.

natebeauregard commented on yarn-project/p2p/src/mem_pools/tx_pool/aztec_kv_tx_pool.ts:
```diff
@@ -106,12 +138,16 @@ export class AztecKVTxPool implements TxPool {
         const tx = await this.getTxByHash(hash);
         if (tx) {
-          await this.#pendingTxPriorityToHash.set(getPendingTxPriority(tx), key);
+          await this.addPendingTxIndices(tx, key);
```
I had the `#pendingTxSize` updates in `addPendingTxIndices` originally, however I noticed that my unit test for evicting low priority txs was failing because the cached pool size (`#pendingTxSize`) wasn't increasing as expected. 

While debugging, I moved the `#pendingTxSize` updates out of the `Promise.all` statement inside of the DB transaction and saw that the pending tx size updated as expected. Additionally, we'll only write to the DB once per `addTxs` call now for `#pendingTxSize` instead of incrementing for each individual tx.

natebeauregard committed:
chore: address perf bottlenecks and update unit test mocks

natebeauregard committed:
chore: additional perf updates and added fee payer unit test for reorgs

natebeauregard commented:
@spalladino I addressed your comments and added some perf improvements based on my benchmarking earlier today and your earlier suggestions.

I'm happy to keep the perf updates in a separate branch if we want the base implementation to merge in this PR and then any perf updates in a follow-up PR after this runs on master for a little bit, let me know if you think I should leave them in or take them out for the initial PR.

Also, would you mind adding the `ci-external` label to this PR so I can kick off a CI run?

spalladino commented on yarn-project/p2p/src/mem_pools/tx_pool/aztec_kv_tx_pool.ts:
```diff
@@ -479,26 +532,30 @@ export class AztecKVTxPool implements TxPool {
       }
-      const [index] = await archiveCache.getArchiveIndices([headerHash]);
+
+      const archive = headerHash ? Fr.fromString(headerHash) : await tx.data.constants.historicalHeader.hash();
```
Can `headerHash` be undefined here? I'd expect to always have a value, since we're iterating over the key-value pairs. It'd also allow us to not have to call `getPendingTx`, right?

natebeauregard replied:
Yeah good point, `headerHash` should always be defined. I'll remove that undefined check for `headerHash`, however I think we still need the `getPendingTxByHash` call to do the fee payer balance validations with `gasTxValidator.validateTxFee(tx)`.

spalladino reviewed (approved):
Looks good, thanks for the work here Nate! Optimizations look good, no need to split them out to another PR. As a safeguard, in a future PR, WDYT about having a process that runs every now and then (could be every 10 mined blocks, or every hour, or sth like that) and checks the consistency of all indices? I'm worried about a corner case we may have missed, so having a process that will alert us loudly in the logs if something's off will help identify these issues.

natebeauregard committed:
fix: prevent race condition during reorgs

natebeauregard committed:
Merge branch 'master' into nb/mempool-limiting

natebeauregard commented:
> Looks good, thanks for the work here Nate! Optimizations look good, no need to split them out to another PR. As a safeguard, in a future PR, WDYT about having a process that runs every now and then (could be every 10 mined blocks, or every hour, or sth like that) and checks the consistency of all indices? I'm worried about a corner case we may have missed, so having a process that will alert us loudly in the logs if something's off will help identify these issues.

Great idea, created https://github.com/AztecProtocol/aztec-packages/issues/13393 to eventually add the process
